feat(core): handle slices impossible to save on figma

gabrieleAngius opened the pull request:
# Proposed changes

Added an action to generate the theme page (with some preset of slices) and a Box component (which contains all the variants). Added also a Sync action to update the existing Box variants according to the current styles on the slices frames (and add some new if any)

wip #7

## Types of changes

- [x] ✨ New feature (non-breaking change which adds functionality or enhancements)
- [ ] 🐛 Bugfix (non-breaking change which fixes an issue)
- [ ] 💥 Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [x] ✅ Tests (added tests for an existing feature)
- [ ] 📚 Chore / Documentation Update (if none of the other choices apply)
- [ ] 🙌 Other (please, write a clear and concise description of the proposal in the section above)

## Checklist

- [x] :scissors: I have ensured that the PR is concise and broken down if needed
- [x] 👀 I have read the [README](../README.md) doc
- [x] 🧪 I have added tests that prove my fix is effective or that my feature works
- [x] 📚 I have added necessary documentation (if appropriate)
- [x] 🔀 Any dependent changes have been merged and published in downstream modules

Commit:
feat(core): handle slices impossible to save on figma

added an action to generate the theme page (with some preset of slices) and a Box component (which
contains all the variants). Added also a Sync action to update the existing Box variants according
to the current styles on the slices frames (and add some new if any)

wip #7

Commit:
feat(core): improved copy on a utils test

Commit:
fix(core): refactoring of getVariantCombinations function

lucacacciarru commented on src/plugin/pluginControllers.ts:
```diff
@@ -13,6 +24,97 @@ export const controllers: Record<PluginActionTypes, Resolver> = {
+
+  'generate-theme-page': () => {
+    const page = figma.createPage();
+    page.name = 'Theme';
```
If the user already has a page named in this way, can Figma create it? Maybe we should call it something like "Morfeo theme" or something like that

gabrieleAngius replied:
Yes, if the user already have a page named "Theme" the plugin will create it anyway. But indeed that's bad since this name should be unique. I think I'll add this name to a constant, and change to "#Morfeo theme". It's also nice to add a check to prevent the plugin creating a duplication of this page

lucacacciarru replied:
So, we can name this page "Theme" if there are no other pages with this name. Otherwise "#Morfeo theme"

gabrieleAngius replied:
Since we need to use the name as reference when sync slices&components, we should be sure what is the name, so:

- the name will be ever the same
- the user will be notified if a page with such name already exist (and the plugin will not create a new one)

Just committed those changes, check it out

lucacacciarru replied:
Well done ❤️

lucacacciarru commented on src/plugin/page-generation/utils.ts:
```diff
@@ -0,0 +1,236 @@
+  const existentBorderWidthSlices: Record<string, number> = {};
+
+  borderWidthsFrame.children.map((borderWidthSlice) => {
+    if (borderWidthSlice.type !== 'LINE') {
```
I don't think this is important now, but we can consider allowing the user to also use the rectangle to set this design token

gabrieleAngius replied:
Not sure about that, it'll add some complexity (a few checks to understand what type is the slice and how to behave for each case). But even if from UX perspective could be good (less restrictive rules), IMO for now we should decide just a single way and enforce it.

Looks like this topic should be discussed, could you open a dedicated issue?

lucacacciarru replied:
Sounds good, I'm gonna open the issue to discuss about it

Commit:
fix(core): changed the name of Theme and added a check to prevent creating it if already exist

changed name of Theme to "#Morfeo theme", extracted it to a constant and added a check to prevent
creating it in case it already exist. Added some tests to cover both cases

wip #7
